### PR TITLE
Auto-complete Dataset

### PR DESCRIPTION
**Pull request by charleslparker**
This is a little script that takes a dataset that may have missing data, trains a model for each missing field, and uses the model to fill in the missing data.  It also inserts empty strings for missing text/items fields and removes non-preferred fields in the bargain.  Finally, for each missing field that it had to "complete", it adds a column indicating whether or not the field was missing in the original dataset.

This may be useful for the upcoming Laminar, as DBNs don't deal with missing data in a natural way.  Also for logistic regression where we don't feel like our built-in options for imputation are enough (/cc @cheesinglee).

This was inspired by the talks that Poul and I gave in Valencia, as well as a conversation I had afterwards with someone who asked for it specifically.

@jaor - Is there a reason that Poul and FM aren't in this group?  I'd like to cc them but I guess I can't unless they are in the whizzml "organization".

### CODE REVIEW

**Comment by jaor** [comment]
> @jaor - Is there a reason that Poul and FM aren't in this group? I'd
> like to cc them but I guess I can't unless they are in the whizzml
> "organization".

@aficionado hasn't accepted his invitation yet (sent months ago).  i
just sent one to @petersen-poul.

**Comment by whizzmler** [comment]
> @jaor https://github.com/jaor - Is there a reason that Poul and FM
> aren't in this group? I'd like to cc them but I guess I can't unless they
> are in the whizzml "organization".

  ? I got this...

-poul

**Comment by charleslparker** [comment]
Poul - it looks like you got it as the user @whizzmler, but @petersen-poul doesn't belong to the group.  In any case, I thought not belonging to the group meant that @-mentions wouldn't parse, but I guess they just don't auto-complete and are still communicated.

File: repair-missing/metadata.json
Comment (akashenfelter) [inline repair-missing/metadata.json]: Missed a word: "Can be **used** to complete future datasets of the same format."

File: repair-missing/script.whizzml
Comment (akashenfelter) [inline repair-missing/script.whizzml:4]: This may be a newbie question, but why did you go with a loop here instead of a filter?

File: repair-missing/script.whizzml
Comment (akashenfelter) [inline repair-missing/script.whizzml]: But the `datetime` optype isn't necessarily present in all datasets with generated data, so those datasets won't get flattened and the batchprediction will still mess up their field ids.  Is there something I'm missing?

File: repair-missing/readme.md
Comment (akashenfelter) [inline repair-missing/readme.md]: Maybe add in a `VERBOSITY` value? But this is wonderful and I am stealing your language for my test readmes.

File: repair-missing/script.whizzml
Comment (akashenfelter) [inline repair-missing/script.whizzml:38]: And another newbie question: why are you passing the field id and field descriptions as two separate variables? To save lookup times?

**Review by akashenfelter — CHANGES_REQUESTED** [review]
One typo and a few questions that my be more about my understanding than your code. Thanks for being my first code review!

File: repair-missing/readme.md
Comment (charleslparker) [reply repair-missing/readme.md]: I stole this almost completely from @mmerce 's Brouta feature selection script!  I also stole most of the test script itself so I'm not sure where or how the verbosity value usually works.  I don't see it in `anomaly-benchmarking` either?  Can you point me to this, or am I just missing it?

File: repair-missing/script.whizzml
Comment (charleslparker) [reply repair-missing/script.whizzml:38]: No, I just need both pieces of information in the function.  I don't see a particularly cleaner alternative . . . unless you're suggesting I look up the descriptor from the `train-ds` already sitting there?  In that case, yes, I was just trying to save network call to pull down the dataset resource (you see below that this function is used in a `map`, so the time might be non-trivial depending on the list size).  Are you thinking of something else?

File: repair-missing/script.whizzml
Comment (charleslparker) [reply repair-missing/script.whizzml:4]: Not at all a newbie question!  I copy-pasted this from a time before we built in `filter`, which would indeed be better here.  Though in this case (selecting one item), I prefer `some`.  I'll change it.

File: repair-missing/script.whizzml
Comment (charleslparker) [reply repair-missing/script.whizzml]: There's almost certainly something _I'm_ missing.  I thought datetimes were the only time we were generating fields, but you're saying there are others?  What are they?

File: repair-missing/metadata.json
Comment (charleslparker) [reply repair-missing/metadata.json]: Indeed

**Comment by charleslparker** [comment]
A few return questions and a few fixes.  Also, since it looks like @jaor is going to push out the generated fields fix before the next prod release, let's wait to merge this until after the release, and I'll just dump my (apparently broken) hack around that.

File: repair-missing/script.whizzml
Comment (akashenfelter) [reply repair-missing/script.whizzml]: I'm not sure if it counts as "generated", but my own troubles with the batchprediction name switcharoo are with an extended dataset that doesn't have a `datetime`. For example, take a dataset with fields 000000 through 00000e, and use a flatline expression to add a new field. By default, it will have id 10000f. Use that as the objective field of a logistic regression, and get the output dataset from a batchprediction, and now it's 00000f. 

But yes, jao seems to have fixed this! Wooo!

File: repair-missing/script.whizzml
Comment (akashenfelter) [reply repair-missing/script.whizzml:4]: Ah yes, `some` is even better.

File: repair-missing/script.whizzml
Comment (akashenfelter) [reply repair-missing/script.whizzml:38]: That's what I was thinking, looking it up from the the `train-ds`. Which, now that you point it out, is a lot of calls. This is something I'm learning to pay attention to when coding, thanks for clarifying!

File: repair-missing/readme.md
Comment (akashenfelter) [reply repair-missing/readme.md]: This messed me up when I learned about running tests. When you run `./test.sh`, you don't get much output. Oftentimes, just an error message if the resources don't clean up in a neat order. With `VERBOSITY=1 ./test.sh`, you see all the lovely log messages, and with `VERBOSITY=2 ./test.sh` you'll see every little thing bigmler is doing. So I ran `./test.sh` and thought my test was bad, but with `VERBOSITY=1 ./test.sh` I could see it was running fine.

And I should really go improve that in `anomaly-benchmarking` too.

**Comment by akashenfelter** [comment]
Sounds great! I'll merge the changes then.

**Comment by jaor** [comment]
re verbosity, that when everything goes well there's not ouput at all is
_by design_: i only want to see output in case of errors (the
unfortunate fact the there's still a bit of output when there are no
errors is, well, unfortunate, and we should check and see if bigmler can
clean up resources without complaining).

so please, keep things quiet when VERBOSITY is 0 :)

**Comment by akashenfelter** [comment]
Yes, absolutely! Maybe it's just a personal quirk (unfamiliarity with running tests?) that I always set VERBOSITY to 1 and like to see the message that everything went ok. There is definite value in having the default be silence when everything went right. No news is good news. :)

File: repair-missing/script.whizzml
Comment (charleslparker) [reply repair-missing/script.whizzml]: Ah, yes.  Now that you mention this I remember seeing it before.

File: repair-missing/readme.md
Comment (charleslparker) [reply repair-missing/readme.md]: Wow - I was thinking I had to make some changes to make this happen, but I just ran `VERBOSITY=2 ./test.sh` and auto-magically got the extra output!  I'm still not sure exactly how this happened but something about gift horses and mouths applies here.  I added a bit in the readme to let people know about the option.

**Comment by charleslparker** [comment]
> Sounds great! I'll merge the changes then.

Please wait until I give one final go-ahead; once it's tested I'll remove the unncessary functions and let you know when I'm done.  Thanks!

**Comment by mmerce** [comment]
@akashenfelter, we're alike. When I run the tests I expect them to print some output to be sure of what they're doing. This must be due to my controller personality :-). However, @jaor is usually only interested in them if they fail, so he set this VERBOSITY behaviour.

In BigMLer, --verbosity 1 is the default and shows the output and --verbosity 0 is silent. You can even see the requests and response for the http calls if you add --debug at the end. If that's something you could need, we could even add another VERBOSITY code for that.

**Comment by charleslparker** [comment]
Okay, just tested against labs without flattening and everything seems to work.  This is ready to be merged once prod has the batch prediction fix in place.

**Comment by akashenfelter** [comment]
Wintermute is updated, merging.

**Pull request merged**
